### PR TITLE
doe_pci_cfg: add does status error handling

### PR DESCRIPTION
**Pull request by twilfredo**
This adds support to checking the DOE error bit, if set by the device, spdm-utils will issue a DOE abort and terminate the session.

### CODE REVIEW

File: src/doe_pci_cfg.rs [hunk 1]
```diff
@@ -135,7 +145,12 @@ unsafe extern "C" fn doe_pci_cfg_receive_message(
     info!("Receiving message");
     let pcie_ids = PCIE_IDENTIFIERS.get().unwrap();
     let (pacc, device, doe_offset) = get_pcie_dev(pcie_ids.vid, pcie_ids.devid).unwrap();
-    doe_wait_status_dor(device, doe_offset).unwrap();
+    _ = doe_wait_status_dor(device, doe_offset).map_err(|e| match e {
```
Comment (alistair23) [inline src/doe_pci_cfg.rs:148]: Why do we have the `_ =`?

Won't that contain the `LIBSPDM_STATUS_ERROR_PEER` as we are in the closure

**Review by alistair23 — CHANGES_REQUESTED** [review]

File: src/doe_pci_cfg.rs (hunk 1 above)
Comment (twilfredo) [reply src/doe_pci_cfg.rs:148]: Closure returns from the function on error with `?` the `_` is for the `Ok(())`

Comment (twilfredo) [reply src/doe_pci_cfg.rs:148]: oh wait it doesn't

Comment (twilfredo) [reply src/doe_pci_cfg.rs]: good catch! fixed

File: src/doe_pci_cfg.rs [hunk 2]
```diff
@@ -76,7 +81,14 @@ unsafe extern "C" fn doe_pci_cfg_send_message(
     info!("Sending message {msg_buf:x?}");
     let pcie_ids = PCIE_IDENTIFIERS.get().unwrap();
     let (pacc, device, doe_offset) = get_pcie_dev(pcie_ids.vid, pcie_ids.devid).unwrap();
-    doe_wait_not_busy(device, doe_offset).unwrap();
+    if let Err(e) = doe_wait_not_busy(device, doe_offset).map_err(|e| match e {
+        DoeStatus::DoeStatusErr => {
+            doe_issue_abort(device, doe_offset, true);
+            return LIBSPDM_STATUS_ERROR_PEER;
```
Comment (alistair23) [inline src/doe_pci_cfg.rs:87]: This isn't returning an `Err()` so the `if let` isn't going to hit it

Comment (alistair23) [reply src/doe_pci_cfg.rs:87]: This should probably be something more like

    if let Err(e) = doe_wait_status_dor(device, doe_offset) {
      match e {
        DoeStatus::DoeStatusErr => {
            doe_issue_abort(device, doe_offset, true);
            return LIBSPDM_STATUS_ERROR_PEER;
        }
    }

Comment (twilfredo) [reply src/doe_pci_cfg.rs:87]: aw yeah thanks! fixed

File: src/doe_pci_cfg.rs [hunk 3]
```diff
@@ -464,20 +493,57 @@ unsafe fn doe_wait_not_busy(device: *mut pci_dev, doe_offset: i32) -> Result<(),
...
+    pci_write_long(
+        device,
+        doe_offset + DOE_CONTROL,
+        doe_control | DOE_CONTROL_ABORT,
+    );
+
+    if block_on_status_err {
+        debug!("Waiting for DOE status error to clear");
+        let mut doe_status = pci_read_long(device, doe_offset + DOE_STATUS);
+        while doe_status & DOE_STATUS_ERR != DOE_STATUS_ERR {
```
Comment (alistair23) [inline src/doe_pci_cfg.rs:540]: Shouldn't this be `==`

File: src/doe_pci_cfg.rs (hunk 2 above)
Comment (alistair23) [reply src/doe_pci_cfg.rs]: Actually, reading https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err again what you originally had also worked!

Sorry, I misunderstood how `map_err()` works. Either is fine then

Comment (twilfredo) [reply src/doe_pci_cfg.rs]: ha okay I was a bit confused

File: src/doe_pci_cfg.rs (hunk 3 above)
Comment (twilfredo) [reply src/doe_pci_cfg.rs]: yeah! my brain is on power saving mode today it seems

File: src/doe_pci_cfg.rs (hunk 2 above)
Comment (alistair23) [reply src/doe_pci_cfg.rs]: You can push back then :)

**Review by alistair23 — APPROVED** [review]

**Pull request merged**